python/cuml/cuml/internals/interop.py

```diff
@@ -0,0 +1,311 @@
+            If one or more hyperparameters are unsupported by the CPU model.
+        """
+        raise NotImplementedError
+
+    def _attrs_from_cpu(self, model) -> dict[str, Any]:
+        """Get attributes to set on ``self`` from a fit CPU model.
+
+        The callers of this method check that ``model`` is of the correct
+        type and has been fit, no need to check that in an implementation.
+
+        The base class handles common metadata attributes, implementations should
+        be sure to call ``super()._attrs_from_cpu()`` to include those in the output.
+
+        Parameters
+        ----------
+        model
+            The CPU model, of the same type as ``self._cpu_class``.
+
+        Returns
+        -------
+        dict
+            A mapping of attributes to set on ``self`` to create a fit GPU model
+            from a fit CPU model.
+
+        Raises
+        ------
+        UnsupportedOnGPU
+            If one or more attributes are unsupported by the GPU model.
+        """
+        out = {}
+        for name in ["n_features_in_", "feature_names_in_"]:
+            try:
+                out[name] = getattr(model, name)
+            except AttributeError:
+                pass
+        return out
+
+    def _attrs_to_cpu(self, model) -> dict[str, Any]:
+        """Get attributes to set on CPU model ``model`` from ``self``.
+
+        The callers of this method check that ``model`` is of the correct
+        type and that ``self`` has been fit, no need to check that in an implementation.
+
+        The base class handles common metadata attributes, implementations should
+        be sure to call ``super()._attrs_to_cpu()`` to include those in the output.
+
+        Parameters
+        ----------
+        model
+            The CPU model, of the same type as ``self._cpu_class``.
+
+        Returns
+        -------
+        dict
+            A mapping of attributes to set on ``model`` to create a fit CPU model
+            from a fit GPU model.
+
+        Raises
+        ------
+        UnsupportedOnCPU
+            If one or more attributes are unsupported by the CPU model.
+        """
+        out = {}
+        for name in ["n_features_in_", "feature_names_in_"]:
+            try:
+                out[name] = getattr(self, name)
+            except AttributeError:
+                pass
+        return out
+
+    def _sync_attrs_to_cpu(self, model) -> None:
+        """Sync any fitted attributes from ``self`` to ``model``.
+
+        Parameters
+        ----------
+        model
+            An instance of ``self._cpu_class``.
+        """
+        if getattr(self, "n_features_in_", None) is None:
+            # GPU model not fitted, nothing to do
+            return
+
+        # XXX: we use this for now to ensure _attrs_to_cpu can rely on
+        # a consistent type for all fitted attributes, rather than
+        # having things potentially vary based on `self.output_type`.
+        with using_output_type("cuml"):
```

Comment:
what does the value `"cuml"` mean here? I had a look at the docs of `using_output_type()` which doesn't mention it, but it does seem to appear in a `if` statement in it.

Reply:
`CumlArrayDescriptor` attributes yield `CumlArray` instances. We want the method to access a consistent attribute type (per the comment above), that seemed the sanest one.